data_transfer/lib/byteflies.py

```diff
@@ -0,0 +1,225 @@
+    serial = __key_by_value(config.byteflies_devices, uuid)
+    return serial
+
+
+def __key_by_value(filename: Path, needle: str) -> Optional[str]:
+    """
+    Helper method to find key in CSV by value (needle)
+    """
+    data = read_csv_from_cache(filename)
+    for item in data:
+        if needle == item["Serial"]:
+            return str(item["Asset Tag"])
+    return None
+
+
+def __get_response(session: requests.Session, url: str) -> Any:
+    """
+    Wrapper method to execute a GET request. Gives a second
+    break to avoid 429 / 502 TooManyRequests (as advised)
+    """
+    # TODO: manage ByteFlies API requests through other means than time.sleep
+    time.sleep(1)
+    response = session.get(url)
+    if code := response.status_code != 200:
+        # when too many requests, we expect 429 or 502
+        if code != 429 and code != 502:
+            # TODO: catch/log exception
+            response.raise_for_status()
+        else:
+            # wait and retry soon-ish
+            pass
+
+    return response.json()
+
+
+def __get_groups(session: requests.Session) -> List[str]:
+    """
+    Returns the list of groupIds associated with a User Account.
+    NOTE: groupIds are hardcoded, so this method is here for posterity
+    """
+    groups = __get_response(session, f"{config.byteflies_api_url}/groups/")
+    group_ids = [g["groupId"] for g in groups]
+    return group_ids
+
+
+def __get_recordings_by_group(
+    session: requests.Session, studysite_id: str, begin_date: str, end_date: str
+) -> Any:
+    """Returns the list of Recordings associated to a Group"""
+    return __get_response(
+        session,
+        f"{config.byteflies_api_url}/groups/{studysite_id}"
+        f"/recordings?begin={begin_date}&end={end_date}",
+    )
+
+
+def __get_recording_by_id(
+    session: requests.Session, studysite_id: str, recording_id: str
+) -> Any:
+    """Returns the details of a particular Recording"""
+    return __get_response(
+        session,
+        f"{config.byteflies_api_url}/groups/{studysite_id}/recordings/{recording_id}/",
+    )
+
+
+def __get_algorithm_uri_by_id(
+    session: requests.Session, studysite_id: str, recording_id: str, algorithm_id: str
+) -> str:
+    """Returns the details of a particular Recording"""
+    payload = __get_response(
+        session,
+        f"{config.byteflies_api_url}/groups/{studysite_id}"
+        f"/recordings/{recording_id}/algorithms/{algorithm_id}",
+    )
+    return str(payload["uri"])
+
+
+def __download_file(
+    session: requests.Session, download_folder: str, url: str, filename: str
+) -> None:
+    """
+    Builds the target filename and starts downloading the file to disk
+    NOTE: Download folder is always /PATIENT_ID/DEVICE_ID
+    """
+    folder = Path(config.storage_vol) / download_folder
+    if not folder.exists():
+        folder.mkdir(parents=True, exist_ok=True)
+    file_path = folder / f"{filename}.csv"
+    response = session.get(url, stream=True)
```

Comment:
Note: when using `Stream=True` we should call this assignment within a [with statement](https://requests.readthedocs.io/en/master/user/advanced/#body-content-workflow) to ensure the contet is closed. That would mean the reading file bit below would be intended within this with similar to how I did it in the latest [dreem branch](https://github.com/ideafast/middleware-services/compare/docker_pipeline#diff-666c9d8f23463c4da86839196dc51f9d3b26e0e36d7a3e05bdf0cce458f13814R139-R140).